File: Assignment/Implementation/Utils/SymbolTable.cs
using Assignment.Abstraction;
using Assignment.Abstraction.Utils;
using Assignment.Implementation.Errors;
using System;
using System.Collections;

namespace Assignment.Implementation.Utils
{
    internal class SymbolTable
    {
        public SymbolTable Parent { get; }
        public Scope Scope { get; }
        private readonly Hashtable _table = new Hashtable();

        public SymbolTable()
        {
            this.Scope = new Scope(Guid.NewGuid());
            this.Parent = null;
        }

        public SymbolTable(Scope scope)
        {
            this.Scope = scope;
        }

        public SymbolTable(Scope scope, SymbolTable parent)
        {
            this.Scope = scope;
            this.Parent = parent;
        }

        public SymbolTable(SymbolTable parent)
        {
            this.Scope = new Scope(Guid.NewGuid());
            this.Parent = parent;
        }

        public void Add(Symbol symbol)
        {
            symbol.Scope = this.Scope;
            _table.Add(symbol.Name, symbol);
        }

        public bool Contains(string name)
        {
            if (_table.ContainsKey(name))
                return true;

            if (this.Parent != null)
                return this.Parent.Contains(name);

            return false;
        }

        public bool TryGet(string name, out Symbol symbol)
        {
            if (_table.ContainsKey(name))
            {
                symbol = (Symbol)_table[name];
                return true;
            }

            if (this.Parent != null)
            {
                return this.Parent.TryGet(name, out symbol);
            }

            symbol = default(Symbol);
            return false;
        }
    }
}

File: src/main/java/ce305/implementation/utils/SymbolTable.java
package ce305.implementation.utils;

import java.util.Hashtable;
import java.util.UUID;

import ce305.abstraction.utils.Scope;
import ce305.abstraction.utils.Symbol;

public class SymbolTable {
    public SymbolTable parent;
    public Scope scope;
    private Hashtable<String, Symbol> _table = new Hashtable<>();

    public SymbolTable()
    {
        this.scope = new Scope(UUID.randomUUID());
        this.parent = null;
    }

    public SymbolTable(Scope scope)
    {
        this.scope = scope;
    }

    public SymbolTable(Scope scope, SymbolTable parent)
    {
        this.scope = scope;
        this.parent = parent;
    }

    public SymbolTable(SymbolTable parent)
    {
        this.scope = new Scope(UUID.randomUUID());
        this.parent = parent;
    }

    public void add(Symbol symbol) {
        symbol.scope = this.scope;
        _table.put(symbol.name, symbol);
    }

    public boolean contains(String name) {
        if (_table.containsKey(name)) {
            return true;
        }

        if (this.parent != null) {
            return this.parent.contains(name);
        }

        return false;
    }

    public Symbol get(String name) {
        if (_table.containsKey(name)) {
            return _table.get(name);
        }

        if (this.parent != null && this.parent.contains(name)) {
            return this.parent.get(name);
        }

        return null;
    }
}